core-rust/state-manager/src/staging/result.rs

```diff
@@ -180,24 +180,34 @@ impl ProcessedCommitResult {
         );
         let state_hash_tree_diff =
             Self::compute_state_tree_update(store, parent_state_version, &database_updates);
-        let transaction_tree_diff = Self::compute_accu_tree_update::<S, TransactionTreeHash>(
-            store,
-            epoch_transaction_identifiers.state_version,
-            epoch_transaction_identifiers.transaction_hash,
-            parent_state_version,
-            TransactionTreeHash::from(ledger_transaction_hash.into_hash()),
-        );
+
+        let epoch_accu_trees =
+            EpochAwareAccuTreeFactory::new(epoch_identifiers.state_version, parent_state_version);
+
+        let transaction_tree_diff = {
+            let mut collector = CollectingAccuTreeStore::new(store);
```

Comment:
To be honest, I find all of this indirection, with Factories, Stores, CollectingStores and Trees a little bit confusing. But I take it it's all been tested and is okay :).

Not something I'd suggest to change as part of this PR at any rate.

Reply:
I see what you mean...

I imagine that a true rust developer would have a method like:
`current_tree.compute_next_tree_diff_after_adding(leaf_hashes: Vec<N>) -> AccuTreeDiff<N>`

But OFC a "straight outta java" guy had to make all this needlessly flexible, via 7 interfaces and a `AccuTreeProxyStoreManagerFactoryCreator` implementation.
In java this is less INconvenient (and often arguably LESS complicated) because there are no "borrow checker rules" (I mean it).

In our use-case - it turns out - we pay the complexity price, but we never need the flexibility (i.e. we always kinda collect the result diff and pass it down; we never write to the accu store directly). So it seems like the abstractions could be cut down considerably. TBH I don't find these particular abstractions overly-complicated (in general, splitting a single domain into [a store, a factory, a collector, whatever] is valid to me) and I wouldn't remove them just yet.

BUT it is certainly true that we can hide some of this complexity from the usual use-case that we have! So I now added a `epoch_accu_tree.compute_tree_diff(new_leaf_hashes)` convenience method 👍

Reply:
I wouldn't say they're necessarily bad, but just from doing code reviews on low sleep, I haven't spent enough time with them enough to "get" how their abstractions differ / what they are each responsible for 👍.

Happy to keep them :)

Reply:
As for imagining "what an overly-exuberant Rust developer might do"  - personally I'd probably go the trait route - have an `EpochAwareAccumulatorTree` trait with a `Hash: IsHash` associated type, and then just put most of the code in the trait methods :D. But I'm obsessed with traits after all my work on SBOR, traits and transaction versioning :D